### PR TITLE
vendor libp2p with gx

### PR DESCRIPTION
**Pull request by whyrusleeping**
The biggest part of this PR will be making sure everything is still easy to dev with

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 27b7e7f45fb03e3e538b9bcb3a9f841878803761
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: 7b5372eee521c1e291c94af2f550d62e53a32088
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 27b7e7f45fb03e3e538b9bcb3a9f841878803761
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: 7b5372eee521c1e291c94af2f550d62e53a32088
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: 06d0943102de3e3d605acd440f720502c309f340
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 27b7e7f45fb03e3e538b9bcb3a9f841878803761
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: 7b5372eee521c1e291c94af2f550d62e53a32088
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: 06d0943102de3e3d605acd440f720502c309f340
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by whyrusleeping** [comment]
oh my god gitcop, go home youre so dumb

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 27b7e7f45fb03e3e538b9bcb3a9f841878803761
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: 7b5372eee521c1e291c94af2f550d62e53a32088
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: 06d0943102de3e3d605acd440f720502c309f340
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by rht** [comment]
(this should take <=~O(1day) the longest)

**Comment by diasdavid** [comment]
image

**Comment by diasdavid** [comment]
Seal of Approval :D (//cc @RichardLitt)

**Comment by whyrusleeping** [comment]
i still need to fix the tests so they know what the hell is going on

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 27b7e7f45fb03e3e538b9bcb3a9f841878803761
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: 7b5372eee521c1e291c94af2f550d62e53a32088
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: 06d0943102de3e3d605acd440f720502c309f340
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by whyrusleeping** [comment]
now i'm running into docker issues... id *like* to copy $GOPATH/src/gx into the build context, but i dont see any easy way to do that. Its looking like i'll have to install gx on the docker image and run an install... which is annoying because i've already done it once during the CI test runs...

cc @lgierth

**Comment by lgierth** [comment]
Hey @Luzifer please take a look at my docker-related commits here. It's a bit more involved now, we can tune it in another PR -- this here is just to get gx integration going.

**Comment by lgierth** [comment]
@whyrusleeping it's good now

**Comment by whyrusleeping** [comment]
oh my god look those tests are green o.o

**Comment by whyrusleeping** [comment]
@RichardLitt can you take a look at the usability here? Maybe help me out making the readme clear on how this works (all anyone should have to do is run `make install`)

**Comment by whyrusleeping** [comment]
i'll merge this in the morning, we can always improve the readme in a followup PR

**Comment by chriscool** [comment]
In the sharness tests some work has been done to remove any dependency on wget and to use curl instead. It would be nice here to use curl as much as possible instead of wget.

**Comment by lgierth** [comment]
@chriscool wget is being used *inside* the container. I tried curl first but it doesn't have the retrying options I was looking for (retrying unconditionally).

**Comment by rht** [comment]
(in other situation, `ipfs get`/`ipget` was being used in place of `curl` since the latter doesn't do recursive download )

**Comment by RichardLitt** [comment]
@whyrusleeping I opened github.com/ipfs/go-ipfs/pull/2302 to merge into this branch. Is that what you wanted?

File: assets/assets_test.go [hunk 1]
```diff
@@ -9,10 +9,12 @@ import (
 
 // TestEmbeddedDocs makes sure we don't forget to regenerate after documentation change
 func TestEmbeddedDocs(t *testing.T) {
+	t.Skip("skipping for now, paths are hard")
```
Comment (noffle) [inline assets/assets_test.go]: I'm not really familiar with `assets/`: what is this test supposed to do and why are we now skipping it?

**Comment by noffle** [comment]
Nice work @whyrusleeping. It's been cool watching `gx` go from proof of concept to powering go-ipfs.

Some usability feedback:

1. `make` in `cmd/ipfs/` doesn't work: `gx` looks for `package.json`, which of course is back in the root directory.
2. `make` is now very expensive. Every invocation does `go get -u gx`, `go get -u gx-go`, and `gx install`, even when you're up to date. This makes my builds internet dependent, and significantly slower -- adds ~43 seconds. Other folks might have different workflows, but I rely on `make` for development; I'd be really keen on keeping it fast. Maybe having dummy files we `touch` on `go get` and `gx install` would work well with `make`?
3. More makefile stuff: right now only `install` depends on `deps`, but I think `build` targets should rely on `deps` instead -- and then we can have `install` rely on `build`. Does that make sense?


---

Kudos @RichardLitt for the nice clean-up to the README; it's an entry point for so many people who look at IPFS. :+1:

File: assets/assets_test.go (hunk 1 above)
Comment (whyrusleeping) [inline assets/assets_test.go]: the test is testing that the string constants in this file match the files stored in git. Except now those files are stored in gx (and we arent storing the gx deps in git). I can probably add a bit of logic to search for the installed package either locally or globally. Good catch

**Comment by whyrusleeping** [comment]
@noffle good catch on the `cmd/ipfs` makefile, missed that one. As for make being expensive, I do want to ensure that anyone building ipfs as per the listed build instructions has their deps up to date. Building without the makefile is quite easy: `go install ./cmd/ipfs` or `go build ./cmd/ipfs`. We could have a `make fast-install` or `make fast-build` type target for your usecase if you don't want to type the go commands.

having install depend on build depend on deps is a good idea. changing that now.

**Comment by noffle** [comment]
> As for make being expensive, I do want to ensure that anyone building ipfs as per the listed build instructions has their deps up to date.

Agreed. Ideally we'd have versions of `gx` and `gx-go` pinned so we could skip these heavy steps. I'll think about the problem some more, but yeah: it doesn't make sense to hold up this PR because of it.

**Comment by noffle** [comment]
:horse: LGTM

**Comment by rht** [comment]
Somehow missed the `go get -u` in the `make install`. This breaks the immutability of the build of `go-ipfs` of a specific commit (with `gx` remains as being mutable).

**Comment by noffle** [comment]
@rht: yeah +1. I'd like to see a means of retrieving a pinned version of `gx` and `gx-go` as well. Maybe via `ipget` or just `wget`ing the public gateway?

**Comment by rht** [comment]
It is simplest to just bootstrap `gx`, `gx-go` in `vendor/` (this is the closest thing to hash/version pinning in go anyway, albeit implicit). That way, the deps that are fetched by `gx` are still pinned by their hashes or versions.

**Pull request merged**